src/url-handling.js

```diff
@@ -70,6 +62,74 @@ type UrlObject = {
 
 type Query = BaseQuery | CallTreeQuery | MarkersQuery | StackChartQuery;
 
+/**
+ * This function sets up the relationship between the History API, and Redux Store.
+ *
+ * The URL holds the minimal serializable string that represents the current state of
+ * the application. Anything that needs to be serialized to the URL should be stored
+ * in the UrlState. It should be trivial to go between the UrlState and location string.
+ *
+ *    UrlState => string
+ *    string => UrlState
+ *
+ * When the UrlState changes it is serialized back into a location string, and the URL
+ * on the pages is updated. When the location string is manually changed, the string is
+ * deserialized back into a UrlState. However, during browser navigation events, the
+ * UrlState will be pulled from the history state. So while navigating the UrlState
+ * can be pulled out without having to deserialize the location string.
+ */
+export function setupUrlHandling(store: Store) {
+  const { getState } = store;
+
+  // Pull the state out of window.history.state, or deserialize the window.location. This
+  // dispatches the new state to the store so that it can update appropriately.
+  updateStoreFromBrowserEvent(store);
+
+  // Replace the state in history, and update the url after the initial processing just
+  // in case there is any variation.
+  const initialUrlState = getUrlState(getState());
+  window.history.replaceState(
+    initialUrlState,
+    document.title,
+    urlStateToLocationString(initialUrlState)
+  );
+
+  // Send the initial analytics report for a pageview.
+  const dataSource = getDataSource(getState());
+  sendAnalytics({
+    hitType: 'pageview',
+    page: dataSource === 'none' ? 'home' : getSelectedTab(getState()),
+  });
+  sendAnalytics({
+    hitType: 'event',
+    eventCategory: 'datasource',
+    eventAction: dataSource,
+  });
+
+  // Listen for history events, and update the UrlState when those happen.
+  window.addEventListener('popstate', () => updateStoreFromBrowserEvent(store));
+
+  // Finally when the store's UrlState changes, check for differences and update the
+  // URL as needed. Any change to the UrlState should result in a change to the URL,
+  // and a history event pushed on.
+  let oldUrlState = initialUrlState;
+  store.subscribe(() => {
+    const newUrlState = getUrlState(getState());
+
+    // Only update the URL when there are differences.
+    if (!shallowEqual(oldUrlState, newUrlState)) {
+      const oldUrl = window.location.pathname + window.location.search;
+      const newUrl = urlStateToLocationString(newUrlState);
+      if (oldUrl !== newUrl) {
+        window.history.pushState(newUrlState, document.title, newUrl);
+      }
```

Comment:
My concern compared to my "actors" way is we may be running expansive code (concerned about `history.pushState` mainly) too much.  Also I think we mix responsibilities.

1. performance:
`subscribe` is called everytime the state change. Compared to doing this in a react component where (I think) the rendering phase is throttled (is it?). Basically I think we don't want to update the location and the history state too often -- maybe we could throttle it ourselves.

2. mixing responsibilities:
In my "actor" patch, like you do I subscribed to the store, but only dispatched an action when the url state changed (see https://github.com/devtools-html/perf.html/pull/537/files#diff-ecc5e5ed8ddbf07abeebb4d6dde28756R13). At the time, my main concern was to reset the saved "short url" at that moment.

Then UrlManager's responsibility was still the same: actually update the browser UI, as it's the job of React Components.